daemon/images/image_pull.go

```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
 	}
 	return repository, confirmedV2, lastError
 }
+
+type contentStoreForPull struct {
+	distribution.ContentStore
+	leases leases.Manager
+
+	mu       sync.Mutex
+	digested []digest.Digest
+}
+
+func (c *contentStoreForPull) addDigested(dgst digest.Digest) {
+	c.mu.Lock()
+	c.digested = append(c.digested, dgst)
+	c.mu.Unlock()
+}
+
+func (c *contentStoreForPull) getDigested() []digest.Digest {
+	c.mu.Lock()
+	digested := make([]digest.Digest, len(c.digested))
+	copy(digested, c.digested)
+	c.mu.Unlock()
+	return digested
+}
+
+func (c *contentStoreForPull) Writer(ctx context.Context, opts ...content.WriterOpt) (content.Writer, error) {
+	w, err := c.ContentStore.Writer(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &contentWriter{
+		cs:     c,
+		Writer: w,
+	}, nil
+}
+
+type contentWriter struct {
+	cs *contentStoreForPull
+	content.Writer
+}
+
+func (w *contentWriter) Commit(ctx context.Context, size int64, expected digest.Digest, opts ...content.Opt) error {
+	if err := w.Writer.Commit(ctx, size, expected, opts...); err != nil {
+		return err
+	}
+
+	w.cs.addDigested(expected)
+	return nil
+}
+
+func imageKey(id image.ID) string {
+	return "image-" + id.String()
```

Comment:
`moby-image-` ?

This should already be namespaced so not super important but in a follow up we should figure out how to make sure every dockerd is on its own isolated namespace. Eg. if 2 dockerds share same containerd they would now start to delete each others objects.

Reply:
We already have daemon flags b/c of this for integration tests.

Reply:
Yes, but atm it is kind of optional and unlikely to cause damage I believe. Anyway, other than the prefix this is not in the scope of this PR anyway.